### DISCUSSION TITLE
How to Reset YOLOv8n Pretrained Weights to Random Initialization?

### DISCUSSION THREAD

**Discussion by Brayan532** [ultralytics/ultralytics#16537] [Q&A] [answered]
The YOLOv8n model from Ultralytics comes pretrained by default. How can I reset its weights to initialize it with random, untrained weights?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Brayan532, thank you for your question about YOLOv8n! 🚀 This is an automated response, but don't worry, an Ultralytics engineer will assist you soon.

For detailed model instructions, check out the [Docs](https://docs.ultralytics.com) where you can find guidance on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage. 

If this is a 🐛 Bug Report or customization query, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us understand and troubleshoot effectively.

For model training or architecture modifications, you might find our [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) useful. 

Join our vibrant community for discussions and support. Dive into [Discord](https://ultralytics.com/discord) 🎧 for live interactions, explore conversations on [Discourse](https://community.ultralytics.com), or engage over at our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure your `ultralytics` package is upgraded to the latest version including its [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Run YOLOv8 in any up-to-date verified environment with necessary dependencies like [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Refer to the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. Refer to the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. Find instructions in the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests verify YOLOv8's correct operation on various platforms.

Stay tuned for more assistance! 😊

**Comment by glenn-jocher** [comment]
To reset YOLOv8n weights to random initialization, load the model with its YAML configuration file instead of the pretrained weights.

**Reply by Brayan532** [reply]
> To reset YOLOv8n weights to random initialization, load the model with its YAML configuration file instead of the pretrained weights.

Dear Glenn, 
thanks for your reply. Basically, there are two Yaml configuration files:
1. "...\Lib\site-packages\ultralytics\cfg\default.yaml"
2. "....\Lib\site-packages\ultralytics\cfg\models\v8\yolov8.yaml"
As is seen none of them can be used for reset the weights. Indeed, I wanna do it for two purposes: 1. predicting by the Non-trained weights? 2. Training it by Non-trained weights. I would be more than glad if you could assist me in this regard. 

Best wishes.

**Reply by glenn-jocher** [reply]
@Brayan532 to reset the weights, use the `yolov8n.yaml` file located in `...\Lib\site-packages\ultralytics\cfg\models\v8\`. Load the model with this YAML file to initialize with random weights.

**Comment by Brayan532** [comment]
for training I am using model.train(data="data.yaml", epochs=50, imgsz=640, pretrained=False), is it ok?

**Reply by Brayan532** [reply]
I have tried the above approach by using pretrained = False. Nonetheless, it seems this option does not work, since the result of training with or without "pretrained = False" is the same. for your reference please see the first epoch of training as follows that has been the same for the default value of pretrained in the training mode, i.e. True, and pretrained = False. 
 Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/2         0G       1.17      1.387      1.241        216        640: 100%|██████████| 50/50 [08:41<00:00, 10.44s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 25/25 [02:28<00:00,  5.95s/it]
                   all        800       5620      0.688      0.474      0.562      0.405

Have you ever tried to check whether this option works correctly?

**Reply by glenn-jocher** [reply]
@Brayan532 to reset the YOLOv8n model to random weights, you should load the model with its configuration file instead of the pretrained weights. Use `YOLO('yolov8n.yaml')` to initialize the model with random weights. If you have further questions, feel free to ask!

**Reply by Brayan532** [reply]
> @Brayan532 to reset the YOLOv8n model to random weights, you should load the model with its configuration file instead of the pretrained weights. Use `YOLO('yolov8n.yaml')` to initialize the model with random weights. If you have further questions, feel free to ask!

Dear Glenn, 
Thank you so much for your reply. 
Indeed, I have taken your advice with model = YOLO(yaml_path).load('yolov8n.pt') where configuration file is referred to initialise the model with non-trained (random) weights. Surprisingly, no difference is observable. Therefore, may I request you to give me a clue how to set up by a command random weights for this model which may be considered as a solution. Best wishes.

**Reply by glenn-jocher** [reply, marked as answer]
To initialize YOLOv8n with random weights, use `model = YOLO('yolov8n.yaml')` without loading any `.pt` file. This ensures the model starts with untrained weights. Let me know if you need further assistance!

**Reply by Brayan532** [reply]
> To initialize YOLOv8n with random weights, use `model = YOLO('yolov8n.yaml')` without loading any `.pt` file. This ensures the model starts with untrained weights. Let me know if you need further assistance!

Dear glenn, 
First, I sincerely appreciate your response. I also have a few questions that I believe may benefit from your expertise, and I am hopeful you can offer some guidance. Specifically, I have been running YOLOv11n in Google Colab with random weights on a subset of the COCO dataset containing 800 images. Simply put, 

1. yaml_path = "/content/drive/MyDrive/Data/yolo11.yaml",  
2. model = YOLO(yaml_path)
3. result = model.train(data="/content/drive/MyDrive/Data/data.yaml", epochs=40, imgsz=640)

So, I've had 
YOLO11 summary: 319 layers, 2,624,080 parameters, 2,624,064 gradients, 6.6 GFLOPs
Ultralytics 8.3.10 🚀 Python-3.10.12 torch-2.4.1+cu121 CUDA:0 (Tesla T4, 15102MiB)

Then, the mAP50-95 has been observed as for 
epoch 1 -> 0,
epoch 10 -> 7.06e-05, 
epoch 20 -> 0.000193
epoch 28 -> 9.8e-05
epoch 30 -> 0.000133
epoch 40 ->  0.000177

Question 1: Are the current mAP50-95 values in line with what you'd expect in this setting? Should we have anticipated these results given the configuration?

Question 2: You have trained YOLOv11, and it is now available as a pre-trained model with its variants. How many epochs was the model trained for the detection task?

Question 3: I'm training the model from scratch (with random weights as mentioned above). However, before the training begins, I receive the following message:

TensorBoard: Start with 'tensorboard --logdir runs/detect/train2', view at http://localhost:6006/
Freezing layer 'model.23.dfl.conv.weight'

This seems contradictory, as typically no layers are frozen when training from scratch. Could you please share your thoughts on this? Additionally, does this also occur when training a custom YOLOv11 model?

Best wishes.

**Reply by glenn-jocher** [reply]
@Brayan532 thank you for your questions. 

1. The low mAP values are expected when training from scratch, especially with a small dataset. Consider increasing the dataset size or training epochs for better results.
2. YOLOv11 models are typically trained for hundreds of epochs, but specifics can vary.
3. The freezing message is standard and doesn't affect training from scratch; it's part of the model's architecture setup. 

Feel free to reach out if you have more questions!

**Reply by Brayan532** [reply]
I deeply appreciate your swift and clear response.  Warm regards.

**Reply by glenn-jocher** [reply]
You're welcome! If you have any more questions, feel free to ask.
